spec/lib/cms/audit/pages_spec.rb

```diff
@@ -0,0 +1,111 @@
+describe Cms::Audit::Pages do
+  let(:role) { :user }
+  let(:current_user) { create(:user, role: Comfy::Cms::User.roles[role]) }
+  let(:state) { 'published' }
+  let(:en_site) { create(:site) }
+  let(:cy_site) { create(:site, label: 'cy') }
+  let(:page_with_tags) { create(:page_with_multiple_tags, site: en_site, state: state) }
+  let(:page_with_meta) { create(:article_with_metadata, site: cy_site, state: state) }
+  let!(:revision) do
+    create(
+      :revision_with_event,
+      record: page_with_tags,
+      user_id: current_user.id,
+      user_name: current_user.name
+    )
+  end
+  let!(:revision_2) do
+    create(
+      :revision_with_event,
+      record: page_with_meta,
+      user_id: current_user.id,
+      user_name: current_user.name
+    )
+  end
+  let!(:category) { create(:category, site: en_site) }
+  let!(:categorization) do
+    create(
+      :categorization,
+      categorized_id: page_with_tags.id,
+      category: category,
+      categorized_type: 'Comfy::Cms::Page'
+    )
+  end
+
+  before do
+    Timecop.freeze(Time.zone.local(1990))
+  end
+
+  describe 'self.generate' do
+    subject { described_class.generate_report(file: temp_file) }
+
+    let(:temp_file) { Tempfile.new('temp_file.csv') }
+
+    it 'generates a csv file and returns the count of pages added' do
+      expect(subject).to eq 2
+    end
+
+    context 'csv file content' do
+      before do
+        subject
+        @parsed_csv = CSV.read(temp_file)
+      end
+
+      after(:all) do
+        Timecop.return
+      end
+
+      let(:expected_page_1) do
+        [
+          'A page with two tags',
+          '/',
+          'article',
+          category.label,
+          'tag-2,tag-3',
+          nil,
+          nil,
+          nil,
+          'false',
+          'false',
+          'true',
+          '01/01/1990, 00:00',
+          current_user.name,
+          state,
+          'en'
+        ]
+      end
+
+      let(:expected_page_2) do
+        [
+          'Video Page Added',
+          '/video-page-added',
+          'video',
+          '',
+          '',
+          'Just a meta description',
+          'Meta title',
+          'Csv, test',
+          'false',
+          'false',
+          'true',
+          '01/01/1990, 00:00',
+          current_user.name,
+          state,
+          'cy'
+        ]
+      end
+
+      it 'creates a file with correct headers' do
```

Comment:
Totally subjective but in these tests I personally would not rely on a `before` block to execute the `subject under test` and posterior CSV parsing and asserted fields, but go for explicitness  calling both  `audit_generation` (or whatever, I would go for a named subject there), CSV parsing and then the expected fields all together inside each test so I don't need to scroll around to check what is declared for the tests when reading them, especially when `expected_page_`\2` are not reused anywhere so don't need to be extracted from individual tests.

Reply:
I tried it but to be honest it looks messier to me 😄 Especially the array after the `expect`. Unless someone has a very strong opinion against it I would rather leave the variables